### PR TITLE
Printer Status Listener

### LINKED ISSUE #42: Printer status
Request to report back the printer status to the user
- :white_check_mark: Ready
- :warning: Busy
- :warning: Out of paper
- :no_entry: Error
- :no_entry:  Disconnected

This'll be tricky one to implement.  CUPS has some command line utilities for fetching this info and some printers (LPT, COM) may not support bi-directional communication to the spooler, so we'll need a way to soft-fail when the information is not available.

~~If Java and the OS support it, we may be able to leverage `PrinterStateReason` via http://stackoverflow.com/a/5329869/3196753~~ This does not work.

### PR DESCRIPTION
**Pull request by tresf**
Adds a cross-platform printer status listener to QZ Tray.

This won't (yet) tell you if a particular job failed, but does offer general printer status such as printer jam, out of paper, etc.

Supersedes #228
Closes #42

### CODE REVIEW

**Comment by klabarge** [comment]
Windows reporting works pretty well.

Environment:

**Printer:** Epson TM-T88VI Receipt printer
**Driver:** Epson TM-T88VI Receipt5 driver
**OS:** Windows 10 Pro (1803) 64-bit

Results:

* Printer **OK**
   ```
   Thu Sep 13 2018 21:35:05 GMT-0400 (Bolivia Time): OK: Level INFO, StatusCode 0, From EPSON TM-T88VI Receipt5
   ```

* Printer **door open** (returns both `DOOR_OPEN` and `OFFLINE` status)

   ```
   Thu Sep 13 2018 21:38:25 GMT-0400 (Bolivia Time): OFFLINE: Level FATAL, StatusCode 128, From EPSON TM-T88VI Receipt5
   Thu Sep 13 2018 21:38:25 GMT-0400 (Bolivia Time): DOOR_OPEN: Level WARN, StatusCode 4194304, From EPSON TM-T88VI Receipt5
   ```


* Printer **out of paper** (returns both `PAPER_OUT` and `OFFLINE` status)

   ```
   Thu Sep 13 2018 21:40:47 GMT-0400 (Bolivia Time): PAPER_OUT: Level FATAL, StatusCode 16, From EPSON TM-T88VI Receipt5
   Thu Sep 13 2018 21:40:47 GMT-0400 (Bolivia Time): OFFLINE: Level FATAL, StatusCode 128, From EPSON TM-T88VI Receipt5
   ```

* Printer **disconnected** (unplug USB)
   ```
   Thu Sep 13 2018 21:44:04 GMT-0400 (Bolivia Time): NOT_AVAILABLE: Level FATAL, StatusCode 4096, From EPSON TM-T88VI Receipt5
   ```

Note, I tried to test a paper jam, but this printer does not seem to care. I stacked receipt paper until I could barely shut the door, but the printer still tries to spin the feeder and cuts the paper at the end 😅

**Comment by klabarge** [comment]
@Vzor- below are the Java and QZ Tray crash logs that I received on macOS.

[macOS Java crash log.txt](https://github.com/qzind/tray/files/2399017/mac_printerstatus_javacrashlog.txt)

[macOS QZ Tray crash log.txt](https://github.com/qzind/tray/files/2399026/mac_printerstatus_traycrashlog.txt)

Since this crashing occurred, I am unable to reproduce it. I have since re-compiled this branch and reinstalled it. Not sure if that would have anything to do with it.

**Comment by klabarge** [comment]
macOS seems to be working fine. I'm not sure why I cannot reproduce the crashing I originally reported.

I ran tests against an Epson WorkForce Pro 3733 that is connected via WiFi on macOS. I installed the drivers provided by Epson's support website for this model printer. I then added the printer by opening Printers  & Scanners, Add, Selecting the printer in the listing, changing the "Use:" from AirPrint to the printers name. When I left the Printer as "AirPrint", I did not get any status information.

A few notes...
1. When I use "Listen to Current Printer", `printers.getStatus` does not return anything. I must use "Listen to All Printers" in order to get status information
   ```
   [DEBUG] 2018-10-17 23:33:21,093 @ qz.ws.PrintSocketClient:?
	Message: {"call":"printers.startListening","promise":{},"params":{"printerNames":["EPSON WF-3730 Series"]},"timestamp":1539833601093,"uid":"emrgx0"}
   ```
1. If I am listening to a specific printer, I am unable to "Listen to All Printers" without first calling `printers.stopListening`. Does it make sense to automatically close the existing listener and start a new one in this instance, without having to call `printers.stopListening`?

Here are my test results, this feature appears to be working well on macOS.

* Printer **OK**
   ```
   Wed Oct 17 2018 23:13:10 GMT-0400 (GMT-04:00): OK: Level INFO, StatusCode 0, From EPSON_WF_3730_Series, CUPS string none
   ```

* Printer **tray open** (returns `PAPER_PROBLEM` status)

   ```
   Wed Oct 17 2018 23:14:24 GMT-0400 (GMT-04:00): PAPER_PROBLEM: Level FATAL, StatusCode 64, From EPSON_WF_3730_Series, CUPS string input-tray-missing
   ```

* Printer **out of paper** (returns two`PAPER_OUT`)

   ```
   Wed Oct 17 2018 23:25:53 GMT-0400 (GMT-04:00): PAPER_OUT: Level FATAL, StatusCode 16, From EPSON_WF_3730_Series, CUPS string media-empty
   Wed Oct 17 2018 23:25:53 GMT-0400 (GMT-04:00): PAPER_OUT: Level FATAL, StatusCode 16, From EPSON_WF_3730_Series, CUPS string media-needed
   ```

* Printer **disconnected** (power off)
   ```
   Wed Oct 17 2018 23:29:00 GMT-0400 (GMT-04:00): OFFLINE: Level FATAL, StatusCode 128, From EPSON_WF_3730_Series, CUPS string offline
   ```

* Printer in the **middle of print job**
   ```
   Wed Oct 17 2018 23:14:25 GMT-0400 (GMT-04:00): PRINTING: Level INFO, StatusCode 1024, From EPSON_WF_3730_Series, CUPS string cups-waiting-for-job-completed
   ```

**Comment by tresf** [comment]
> When I use "Listen to Current Printer", printers.getStatus does not return anything. I must use "Listen to All Printers" in order to get status information

@Vzor- please see `ShellUtilities. getCupsPrinters()` which has code for fixing the macOS improperly named Java printers to that which CUPS uses.

https://github.com/qzind/tray/blob/911c525e5ad9f0042eaa63fe1158ef4104a1a013/src/qz/utils/ShellUtilities.java#L169-L175 

In regards to the all-or-none behavior, that sounds like a bug but I'll let @Vzor- chime in.

**Comment by Vzor-** [comment]
@klabarge has this last commit fixed the issues with macOS?

File: src/qz/printer/status/StatusMonitor.java [hunk 1]
```diff
@@ -68,6 +69,8 @@ public synchronized static void closeNotificationThreads() {
     }
 
     public synchronized static boolean startListening(SocketConnection connection, JSONArray printerNames) {
+        if (SystemUtilities.isMac()) CupsUtils.convertPrinterNames(printerNames);
```
Comment (bberenz) [inline src/qz/printer/status/StatusMonitor.java]: Is this method/call necessary? Can't we just pull the lookup here and use the values from it directly in this method's loop instead?

File: src/qz/utils/ShellUtilities.java [hunk 2]
```diff
@@ -196,8 +196,8 @@ public static String executeRaw(String[] commandArray) {
                             String[] desc = prop.split(lookFor);
                             if (desc.length > 0) {
                                 // cache the description so we can map it to the actual printer name
-                                descMap.put(entry.getKey(), desc[desc.length - 1].trim());
-                                log.info(entry.getKey() + ": " + desc[desc.length - 1].trim());
+                                descMap.put(desc[desc.length - 1].trim(), entry.getKey());
```
[the remark this answers is not in the document]
Comment (Vzor-) [reply src/qz/utils/ShellUtilities.java]: It simplified the lookup process. With it in the other orientation, loops have to be used to do a reverse lookup.

**Comment by klabarge** [comment]
@Vzor- the latest commit on this branch has broken `printers.find`. I am able to use `printers.find` for an individual printer, but I am not able to list all printers on the system. Below is the stack trace from using "Find all Printers."

```
[DEBUG] 2018-11-29 16:14:33,405 @ qz.ws.PrintSocketClient:?
	Message: {"call":"printers.find","promise":{},"params":{},"timestamp":1543526073404,"uid":"6lp5sq"}
[WARN] 2018-11-29 16:14:33,431 @ qz.ws.PrintSocketClient:?
	Bad signature on request
[ERROR] 2018-11-29 16:14:33,432 @ qz.utils.FileUtilities:?
	Cannot setup file blocked (Shared)
java.io.IOException: No such file or directory
	at java.io.UnixFileSystem.createFileExclusively(Native Method)
	at java.io.File.createNewFile(File.java:1012)
	at qz.utils.FileUtilities.getFile(Unknown Source)
	at qz.auth.Certificate.isBlocked(Unknown Source)
	at qz.ws.PrintSocketClient.allowedFromDialog(Unknown Source)
	at qz.ws.PrintSocketClient.processMessage(Unknown Source)
	at qz.ws.PrintSocketClient.onMessage(Unknown Source)
	at sun.reflect.GeneratedMethodAccessor9.invoke(Unknown Source)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.eclipse.jetty.websocket.common.events.annotated.CallableMethod.call(CallableMethod.java:70)
	at org.eclipse.jetty.websocket.common.events.annotated.OptionalSessionCallableMethod.call(OptionalSessionCallableMethod.java:68)
	at org.eclipse.jetty.websocket.common.events.JettyAnnotatedEventDriver$2.run(JettyAnnotatedEventDriver.java:210)
	at org.eclipse.jetty.util.thread.QueuedThreadPool.runJob(QueuedThreadPool.java:635)
	at org.eclipse.jetty.util.thread.QueuedThreadPool$3.run(QueuedThreadPool.java:555)
	at java.lang.Thread.run(Thread.java:748)
[ERROR] 2018-11-29 16:14:33,460 @ qz.utils.FileUtilities:?
	Cannot setup file blocked (Shared)
java.io.IOException: No such file or directory
	at java.io.UnixFileSystem.createFileExclusively(Native Method)
	at java.io.File.createNewFile(File.java:1012)
	at qz.utils.FileUtilities.getFile(Unknown Source)
	at qz.auth.Certificate.isBlocked(Unknown Source)
	at qz.ui.GatewayDialog.prompt(Unknown Source)
	at qz.common.TrayManager.lambda$showGatewayDialog$3(Unknown Source)
	at java.awt.event.InvocationEvent.dispatch(InvocationEvent.java:301)
	at java.awt.EventQueue.dispatchEventImpl(EventQueue.java:758)
	at java.awt.EventQueue.access$500(EventQueue.java:97)
	at java.awt.EventQueue$3.run(EventQueue.java:709)
	at java.awt.EventQueue$3.run(EventQueue.java:703)
	at java.security.AccessController.doPrivileged(Native Method)
	at java.security.ProtectionDomain$JavaSecurityAccessImpl.doIntersectionPrivilege(ProtectionDomain.java:74)
	at java.awt.EventQueue.dispatchEvent(EventQueue.java:728)
	at java.awt.EventDispatchThread.pumpOneEventForFilters(EventDispatchThread.java:205)
	at java.awt.EventDispatchThread.pumpEventsForFilter(EventDispatchThread.java:116)
	at java.awt.EventDispatchThread.pumpEventsForFilter(EventDispatchThread.java:109)
	at java.awt.WaitDispatchSupport$2.run(WaitDispatchSupport.java:190)
	at java.awt.WaitDispatchSupport$4.run(WaitDispatchSupport.java:235)
	at java.awt.WaitDispatchSupport$4.run(WaitDispatchSupport.java:233)
	at java.security.AccessController.doPrivileged(Native Method)
	at java.awt.WaitDispatchSupport.enter(WaitDispatchSupport.java:233)
	at java.awt.Dialog.show(Dialog.java:1084)
	at java.awt.Component.show(Component.java:1671)
	at java.awt.Component.setVisible(Component.java:1623)
	at java.awt.Window.setVisible(Window.java:1014)
	at java.awt.Dialog.setVisible(Dialog.java:1005)
	at qz.ui.LogDialog.setVisible(Unknown Source)
	at qz.common.TrayManager$4.actionPerformed(Unknown Source)
	at java.awt.MenuItem.processActionEvent(MenuItem.java:669)
	at java.awt.MenuItem.processEvent(MenuItem.java:628)
	at java.awt.MenuComponent.dispatchEventImpl(MenuComponent.java:357)
	at java.awt.MenuComponent.dispatchEvent(MenuComponent.java:345)
	at java.awt.EventQueue.dispatchEventImpl(EventQueue.java:763)
	at java.awt.EventQueue.access$500(EventQueue.java:97)
	at java.awt.EventQueue$3.run(EventQueue.java:709)
	at java.awt.EventQueue$3.run(EventQueue.java:703)
	at java.security.AccessController.doPrivileged(Native Method)
	at java.security.ProtectionDomain$JavaSecurityAccessImpl.doIntersectionPrivilege(ProtectionDomain.java:74)
	at java.security.ProtectionDomain$JavaSecurityAccessImpl.doIntersectionPrivilege(ProtectionDomain.java:84)
	at java.awt.EventQueue$4.run(EventQueue.java:733)
	at java.awt.EventQueue$4.run(EventQueue.java:731)
	at java.security.AccessController.doPrivileged(Native Method)
	at java.security.ProtectionDomain$JavaSecurityAccessImpl.doIntersectionPrivilege(ProtectionDomain.java:74)
	at java.awt.EventQueue.dispatchEvent(EventQueue.java:730)
	at java.awt.EventDispatchThread.pumpOneEventForFilters(EventDispatchThread.java:205)
	at java.awt.EventDispatchThread.pumpEventsForFilter(EventDispatchThread.java:116)
	at java.awt.EventDispatchThread.pumpEventsForHierarchy(EventDispatchThread.java:105)
	at java.awt.EventDispatchThread.pumpEvents(EventDispatchThread.java:101)
	at java.awt.EventDispatchThread.pumpEvents(EventDispatchThread.java:93)
	at java.awt.EventDispatchThread.run(EventDispatchThread.java:82)
[DEBUG] 2018-11-29 16:14:33,471 @ qz.common.TrayManager:?
	Invalid dialog position provided: java.awt.Point[x=0,y=0], we'll center on first monitor instead
[INFO] 2018-11-29 16:14:34,351 @ qz.common.TrayManager:?
	Allowed localhost to access connected printers
[DEBUG] 2018-11-29 16:14:34,364 @ qz.printer.PrintServiceMatcher:?
	Found 4 printers
[DEBUG] 2018-11-29 16:14:34,388 @ qz.utils.ShellUtilities:?
	Executing: [lpstat, -a]
[DEBUG] 2018-11-29 16:14:34,461 @ qz.utils.ShellUtilities:?
	Executing: [lpstat, -l, -p, Zebra_LP2844]
[INFO] 2018-11-29 16:14:34,503 @ qz.utils.ShellUtilities:?
	Zebra LP2844 EPL: Zebra_LP2844
[ERROR] 2018-11-29 16:14:34,503 @ qz.ws.PrintSocketClient:?
	Problem processing message
java.util.ConcurrentModificationException
	at java.util.HashMap$HashIterator.nextNode(HashMap.java:1442)
	at java.util.HashMap$EntryIterator.next(HashMap.java:1476)
	at java.util.HashMap$EntryIterator.next(HashMap.java:1474)
	at qz.utils.ShellUtilities.getCupsPrinters(Unknown Source)
	at qz.utils.PrintingUtilities.getPrinterId(Unknown Source)
	at qz.utils.PrintingUtilities.getDriver(Unknown Source)
	at qz.printer.PrintServiceMatcher.getPrintersJSON(Unknown Source)
	at qz.ws.PrintSocketClient.processMessage(Unknown Source)
	at qz.ws.PrintSocketClient.onMessage(Unknown Source)
	at sun.reflect.GeneratedMethodAccessor9.invoke(Unknown Source)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.eclipse.jetty.websocket.common.events.annotated.CallableMethod.call(CallableMethod.java:70)
	at org.eclipse.jetty.websocket.common.events.annotated.OptionalSessionCallableMethod.call(OptionalSessionCallableMethod.java:68)
	at org.eclipse.jetty.websocket.common.events.JettyAnnotatedEventDriver$2.run(JettyAnnotatedEventDriver.java:210)
	at org.eclipse.jetty.util.thread.QueuedThreadPool.runJob(QueuedThreadPool.java:635)
	at org.eclipse.jetty.util.thread.QueuedThreadPool$3.run(QueuedThreadPool.java:555)
	at java.lang.Thread.run(Thread.java:748)
```

> When I use "Listen to Current Printer", printers.getStatus does not return anything. I must use "Listen to All Printers" in order to get status information

   I also get this `ConcurrentModificationException` error when I use either "Listen to Current Printer" or "Listen to All Printers"

**Comment by klabarge** [comment]
@Vzor- the latest commit fixed the printer listing issues.

> When I use "Listen to Current Printer", printers.getStatus does not return anything. I must use "Listen to All Printers" in order to get status information

   This was fixed 👍 

> If I am listening to a specific printer, I am unable to "Listen to All Printers" without first calling printers.stopListening. Does it make sense to automatically close the existing listener and start a new one in this instance, without having to call printers.stopListening?

   This behavior is still present. Is this okay?

**Comment by Vzor-** [comment]
> > If I am listening to a specific printer, I am unable to "Listen to All Printers" without first calling printers.stopListening. Does it make sense to automatically close the existing listener and start a new one in this instance, without having to call printers.stopListening?
> 
> This behavior is still present. Is this okay?
That behavior is not intended. I did not see that behavior in my earlier tests. I will try again.

**Comment by tresf** [comment]
> That behavior is not intended. I did not see that behavior in my earlier tests. I will try again to reproduce it tonight.

@bberenz changed the behavior in https://github.com/qzind/tray/commit/b2c32ac6f1d2f6c650fded4223a2da4b583cb296#diff-f0516f955a00a62ca82517c7621b73d2L320 probably to be more consistent with our other APIs.  Should we handle this JavaScript side, or have the ability to append files to a listener after it's been started?

**Comment by Vzor-** [comment]
The latest push should allow printers to be appended to the list. The original reason I avoided this was to get around duplicate entries in a MultiMap. I tested the code on windows and did not run into any duplicate listener issue. The same should be true with Ubuntu and macOS though I have not tested them yet.

**Comment by klabarge** [comment]
Tested on macOS.  I can now "Listen to Current Printer", then "Listen to All Printers" without having to call `printers.stopListening`.

However, if I do the below, `printers.getStatus` returns the status of all printers. I would expect `printers.getStatus` to return only the status of the currently selected printer.

1. Listen to All Printers
1. Listen to Current Printer
1. `printers.getStatus`

**Comment by Vzor-** [comment]
> However, if I do the below, `printers.getStatus` returns the status of all printers. I would expect `printers.getStatus` to return only the status of the currently selected printer.

@klabarge You are correct. It is suppose to only show the status of currently registered printers. I will look into it. Do you see the same behavior on Ubuntu?

**Comment by Vzor-** [comment]
@klabarge That should fix it.

**Comment by tresf** [comment]
Some bugs (tested on Mac)

1. Listening for a printer called `"null"` results in all printers being listened.
2. Listening for a printer called `""` results in no errors and no printers being listened.  I've sanitized this through `sample.html` by checking for `NONE` but it should also be handled Java side to harden the API.

I still feel manipulating the JSON data directly on Mac is problematic logic.

**Comment by Vzor-** [comment]
@tresf Seems like a good opportunity to kill two birds with one stone. A printer can be given the name "null" (by a madman) but you cannot name a printer "". For that reason, lets just make "" the listen to all trigger, and nulls/undefined can be turned into "" using a simple falsy catch.

**Comment by tresf** [comment]
>  lets just make "" the listen to all trigger

No, this is inconsistent with the rest of our API.

File: src/qz/printer/status/CupsUtils.java [hunk 3]
```diff
@@ -0,0 +1,240 @@
...
+
+    public static PrinterStatus[] getStatuses(String printerName) {
+        Pointer request = Cups.INSTANCE.ippNewRequest(Cups.INSTANCE.ippOpValue("Get-Printer-Attributes"));
+
+        Cups.INSTANCE.ippAddString(request,
+                                   Cups.INSTANCE.ippTagValue("Operation"),
+                                   Cups.INSTANCE.ippTagValue("uri"),
+                                   "printer-uri",
+                                   "",
+                                   //todo does this need to be sanitized?
```
Comment (tresf) [inline src/qz/printer/status/CupsUtils.java]: Cups doesn't allow spaces or special characters, but it shouldn't hurt to sanitize it anyway incase this behavior changes.

**Comment by Vzor-** [comment]
I tried my best to emulate the behavior of 'printers.find' but it has some quirks; find() finds all printers but find(null) does not. This is because find() passes undefined to the json outbound object, JSON does not technically support undefined, so when stringify is ran, it deletes the property. On the java side of things, we get null when we try to get a string from a non-existent property.  

With status.startListener() things act differently. 'undefined' gets added to an array, and when stringify runs, it wont delete the entire array to remove it, nor will it offset the indices of all other elements to remove it. Instead it censors it to null. When that hits java we pull the 0th element from the array and get the STRING null instead of the object null. Using JSONArray.isNull behaves as we would expect (null = true, "null" = false, 0 = false, "" = false) but as a result we conflate undefined and null.

this issue is pretty minor, but I thought I'd mention it anyway. Also status.startListening("") now returns an error.

TLDR: printers.find() and status.startListening() both target all printers, but status.startListening(null) also listens to all printers where printers.find(null) does not.

**Comment by tresf** [comment]
On macOS mojave, I get a hard crash if I listen to a printer, close the listener and then exit QZ Tray.

Here's the crash logs:
[listener_crash_macos.zip](https://github.com/qzind/tray/files/2751705/listener_crash_macos.zip)

**Comment by tresf** [comment]
The crash seems to be fixed with the following snippet:

```java
NativeLibrary.getInstance("cups").dispose();
```

I've found another hang... if you don't stop listening on macOS, the Java process will hang on exit.

**Comment by tresf** [comment]
> preliminary crash fix 3797024

This fixes the crash on my PC.

**Comment by tresf** [comment]
Clean exit on Ubuntu 16.04 as well.

**Comment by Vzor-** [comment]
the crash was due to a shutdown hook attempting to end a subscription after the cups ipp objects were freed. To fix it I moved the endSubscription into the shutdown hook, and had the ipp objects freed only after it had finished.

**Comment by tresf** [comment]
FYI, Ubuntu throws a benign about 20% of the time when "Stop Listening" is pressed.  After some research, it appears to be a [bug with Jetty 9.2](https://www.eclipse.org/lists/jetty-dev/msg03117.html) and is fixed by upgrading to Jetty 9.3+ which we can do at any time and won't hold up this PR.

```
[WARN] 2019-01-18 15:09:51,226 @ qz.printer.status.CupsStatusServer:?
	Stopping CUPS status server
[WARN] 2019-01-18 15:09:51,249 @ org.eclipse.jetty.server.HttpChannel:396
	/
java.nio.channels.ClosedChannelException
	at org.eclipse.jetty.io.FillInterest.onClose(FillInterest.java:114)
	at org.eclipse.jetty.io.AbstractEndPoint.onClose(AbstractEndPoint.java:110)
	at org.eclipse.jetty.io.AbstractEndPoint.close(AbstractEndPoint.java:116)
	at org.eclipse.jetty.io.ChannelEndPoint.close(ChannelEndPoint.java:113)
	at org.eclipse.jetty.io.SelectChannelEndPoint.close(SelectChannelEndPoint.java:168)
	at org.eclipse.jetty.io.AbstractConnection.close(AbstractConnection.java:225)
	at org.eclipse.jetty.io.SelectorManager$ManagedSelector.closeNoExceptions(SelectorManager.java:723)
	at org.eclipse.jetty.io.SelectorManager$ManagedSelector.access$1100(SelectorManager.java:406)
	at org.eclipse.jetty.io.SelectorManager$ManagedSelector$EndPointCloser.run(SelectorManager.java:1030)
	at org.eclipse.jetty.util.thread.QueuedThreadPool.runJob(QueuedThreadPool.java:635)
	at org.eclipse.jetty.util.thread.QueuedThreadPool$3.run(QueuedThreadPool.java:555)
	at java.lang.Thread.run(Thread.java:748)
```

This branch is ready for final testing and merge.
